File: boot/steps/orchestrator.py
#!/usr/bin/env python3
"""
@format
Bootstrap Orchestrator

Runs the consolidated bootstrap scripts based on node mode.

Usage:
    # Control plane (default):
    python3 orchestrator.py

    # Worker node:
    python3 orchestrator.py --mode worker

    # Dry run (print step list without executing):
    python3 orchestrator.py --dry-run

The orchestrator delegates to control_plane.main() or worker.main()
which handle structured status logging and step sequencing internally.
"""

import argparse
import importlib
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent))
from common import log_info


# =============================================================================
# Module Definitions
# =============================================================================

MODULES = {
    "control-plane": "control_plane",
    "worker": "worker",
}


# =============================================================================
# CLI
# =============================================================================

def main() -> None:
    parser = argparse.ArgumentParser(description="K8s Bootstrap Orchestrator")
    parser.add_argument(
        "--mode",
        choices=["control-plane", "worker"],
        default="control-plane",
        help="Bootstrap mode (default: control-plane)",
    )
    parser.add_argument(
        "--dry-run",
        action="store_true",
        help="Print step list without executing",
    )

    args = parser.parse_args()
    module_name = MODULES[args.mode]

    if args.dry_run:
        log_info(f"DRY RUN — would execute: {module_name}.main()")
        log_info(f"Mode: {args.mode}")
        return

    log_info(f"Orchestrator starting: mode={args.mode}, module={module_name}")
    module = importlib.import_module(module_name)
    module.main()
    log_info("Orchestrator finished")


if __name__ == "__main__":
    main()

File: boot/steps/orchestrator.ts
#!/usr/bin/env tsx
/**
 * @format
 * @module orchestrator
 * Bootstrap Orchestrator — dispatches to {@link control_plane} or {@link worker}
 * based on the `--mode` CLI flag.
 *
 * This thin entry point exists so a single binary (this file) can be invoked by
 * the SSM Automation document for both node types.  The actual bootstrap logic
 * lives in the mode-specific modules to keep concerns separated.
 *
 * @example
 * ```bash
 * # Control-plane bootstrap (default)
 * npx tsx orchestrator.ts
 *
 * # Worker-node bootstrap
 * npx tsx orchestrator.ts --mode worker
 *
 * # Dry-run: print target module without executing
 * npx tsx orchestrator.ts --dry-run
 * npx tsx orchestrator.ts --mode worker --dry-run
 * ```
 */

import { info } from './common.js';

// =============================================================================
// CLI argument parsing
// =============================================================================

/**
 * Parsed CLI arguments for the orchestrator.
 */
interface OrchestratorArgs {
    /** Target bootstrap mode. */
    mode: 'control-plane' | 'worker';
    /** When `true`, logs the target module and exits without executing. */
    dryRun: boolean;
}

/**
 * Parses `process.argv` and returns the orchestrator's runtime arguments.
 *
 * @remarks
 * `--mode` accepts exactly `"control-plane"` or `"worker"`.  Any other value
 * is treated as an error and the process exits with code 1.
 *
 * @returns Parsed {@link OrchestratorArgs}.
 */
const parseArgs = (): OrchestratorArgs => {
    const args = process.argv.slice(2);
    const modeIdx = args.indexOf('--mode');
    const rawMode = modeIdx >= 0 ? args[modeIdx + 1] : 'control-plane';
    const dryRun  = args.includes('--dry-run');

    if (rawMode !== 'control-plane' && rawMode !== 'worker') {
        process.stderr.write(`Invalid --mode: ${rawMode}. Must be 'control-plane' or 'worker'.\n`);
        process.exit(1);
    }

    return { mode: rawMode as OrchestratorArgs['mode'], dryRun };
};

// =============================================================================
// Entry point
// =============================================================================

const { mode, dryRun } = parseArgs();

if (dryRun) {
    const target = mode === 'control-plane' ? 'control_plane.ts' : 'worker.ts';
    info(`DRY RUN — would execute: npx tsx ${target}`);
    info(`Mode: ${mode}`);
    process.exit(0);
}

info(`Orchestrator starting: mode=${mode}`);

if (mode === 'control-plane') {
    const { main } = await import('./control_plane.js');
    await main();
} else {
    const { main } = await import('./worker.js');
    await main();
}
